src/voice/connection.rs

```diff
@@ -71,32 +76,43 @@ impl Connection {
         let url = generate_url(&mut info.endpoint)?;
 
         let mut client = ClientBuilder::from_url(&url).connect_secure(None)?;
+        let mut hello = None;
+        let mut ready = None;
         client.send_json(&payload::build_identify(&info))?;
 
-        let hello = loop {
+        loop {
+            if hello.is_some() && ready.is_some() {
+                break;
+            }
+
             let value = match client.recv_json()? {
                 Some(value) => value,
                 None => continue,
             };
 
             match VoiceEvent::deserialize(value)? {
-                VoiceEvent::Hello(received_hello) => {
-                    break received_hello;
+                VoiceEvent::Ready(r) => {
+                    ready = Some(r);
+                },
+                VoiceEvent::Hello(h) => {
+                    hello = Some(h);
                 },
-                VoiceEvent::Heartbeat(_) => continue,
                 other => {
-                    debug!("[Voice] Expected hello/heartbeat; got: {:?}", other);
+                    debug!("[Voice] Expected ready/hello; got: {:?}", other);
 
                     return Err(Error::Voice(VoiceError::ExpectedHandshake));
                 },
             }
         };
 
-        if !has_valid_mode(&hello.modes) {
+        let hello = hello.unwrap();
+        let ready = ready.unwrap();
```

Comment:
Let's `.expect` these, I don't like unwraps in libraries (even if this should never happen).